### PR TITLE
redo multi project fix to punt objectList<->projectList correlation, …

### PR DESCRIPTION
**Pull request by gabemontero**
…employ more passive project assignment

Fixes https://github.com/openshift/jenkins-client-plugin/issues/130

@openshift/sig-developer-experience ptal

this fix addresses issues introduced by the earlier mulit-project fix for single project envs,
but still maintains the ability to have multi-project creates.

on the bright side, this version simplifies to a decent extent the original multi-project fix.

NOTE: I have obtained agreement from 2 upstream users to validate the fix.  I'm going the attempt to hold merging until we have validation from them, assuming they respond relatively soon.  Also, we do now have a multi-project test case in  the PR extended tests, so we should be good there.

In both upstream users cases, they are hitting timing issues between when selectors with DCs are created, and when they attempt to iterate over pods associated with them with `openshift.withEach`.  While they can hit the issue fairly reliably with their Jenkins files, it has been much less frequent with me (I have seen it like once after many runs).

I also found an issue with the `openshift.union` method ... I plan on creating a new test case for that.

### CODE REVIEW

**Comment by gabemontero** [comment]
Install origin flake 

/retest

**Comment by gabemontero** [comment]
Several confirmations from upstream users have arrived confirming that the manually built plugins from this branch has solved the issue for them.

/hold cancel

@openshift/sig-developer-experience bump ... any takers/folks with bandwidth for a review?

I'll wait until sometime after lunch to see if anyone can.

File: src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy [hunk 1]
```diff
@@ -583,21 +582,13 @@ class OpenShiftDSL implements Serializable {
         }
                 
         Result r = new Result(verb);
-        ArrayList<String> projectNames = new ArrayList<String>();
+        HashMap<String, String> projectNames = null;// = new HashMap<String, String>();
```
Comment (bparees) [inline src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy]: delete the comment?

File: src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy [hunk 2]
```diff
@@ -606,14 +597,21 @@ class OpenShiftDSL implements Serializable {
                         String ns = (String)obj2MapMetadata.get("namespace");
                         if (ns == null || ns.trim().length() == 0)
                             ns = currentContext.getProject();
+                        String kind = obj2Map.get("kind");
+                        String name = obj2MapMetadata.get("name");
                         r = innerObjectDefAction(verb, obj2, userArgs, ns, r);
-                        projectNames.add(ns);
+                        projectNames.put(kind.toLowerCase().trim()+"/"+name, ns);
```
Comment (bparees) [inline src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy]: isn't this going to have collisions if someone is creating a "build/foo" in two namespaces?

**Comment by gabemontero** [comment]
stopped ci job

File: src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy (hunk 1 above)
Comment (gabemontero) [reply src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy]: yep should clean that up

File: src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy (hunk 2 above)
Comment (gabemontero) [reply src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy]: yep you are right

alas, more groovy pain to come as I figure out how to assess :-(

Comment (gabemontero) [reply src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy]: ok I think the change to address is minor .... will test out in the AM

**Comment by gabemontero** [comment]
@bparees comments addressed

**Comment by gabemontero** [comment]
separate commit

**Comment by gabemontero** [comment]
install origin failure .... more instances of https://github.com/openshift/origin/issues/19293

File: src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy [hunk 3]
```diff
@@ -1487,10 +1490,15 @@ class OpenShiftDSL implements Serializable {
...
-                    String project = projectList.get(name);
+                    // if projectList != null then names() return objectList vs. live query
+                    // add index in case same kind/name in diff projects
+                    String project = projectList.get(name+i);
                     if (project != null) {
                         projList = new HashMap<String, String>();
-                        projList.put(name, project);
+                        // we hardcode the suffix to 0 since we are only sending 
+                        // 1 item from the original nameelist down ... hence the 
+                        // suffix is no the index for the first entry in the array
```
Comment (bparees) [inline src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy]: s/nameelist/namelist/
s/no/not? now?/

Comment (gabemontero) [reply src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy]: fixed thanks

**Pull request merged**